Use Templating Component without "templating" service

covex-nn opened the pull request:
I am targeting this branch, because this is backward compatible feature.

See #456

## Changelog

```markdown
### Added
- Added `symfony/asset` and `symfony/templating` dependencies
- Added new service `sonata.templating` for use in place of `templating`

### Changed
- Referencing templates using Twig namespaced syntax

### Removed
- Removed tag `templating.helper` from `sonata.block.templating.helper` service
```

## Subject

This is first step to switching from `Templating Component` to using `Twig` directly. The following two steps are described in #456. This PR will allow to run `SonataBlockBundle` without activating templating extension in `framework`.

If this PR will be accepted, all other sonata bundles, that use SonataBlockBundle, could be updated too: services will have to use `sonata.templating` service instead of `templating` and templates must be referenced using Twig namespaced syntax

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

OskarStark commented on src/Templating/TemplateLocator.php:
```diff
@@ -0,0 +1,21 @@
+use Symfony\Bundle\FrameworkBundle\Templating\Loader\TemplateLocator as FrameworkTemplateLocator;
+
+/**
+ * @deprecated since 3.9, to be removed with 4.0.
```
why? its new

covex-nn replied:
I've created new service `sonata.templating`, to start migrating from `templating` to `twig` service without BC breaks in 3.x branch. `sonata.templating`, like `templating` service, has 3 arguments: `twig`, `sonata.templating.name_parser` and `sonata.templating.locator`. This new class is `sonata.templating.locator` service.

It is a new class, yes. It is needed in 3.x branch to be sure, that everything will be fine after deprecating of `symfony/templating` component. But this new class should not be used by other services, because it will be deleted in master branch (if PR will be merged, of course) accordingly with stage 2 of a plan, described in #456 

So, it is new and deprecated at the same time.

OskarStark commented on src/Templating/TemplateNameParser.php:
```diff
@@ -0,0 +1,21 @@
+use Symfony\Bundle\FrameworkBundle\Templating\TemplateNameParser as FrameworkTemplateNameParser;
+
+/**
+ * @deprecated since 3.9, to be removed with 4.0.
```
why? its new

covex-nn replied:
Same as above. But this class is `sonata.templating.name_parser` service

aLkRicha commented:
Maybe this need to be done in CoreBundle? And then can be reused in other sonata bundles? 
@OskarStark @greg0ire @core23 @jordisala1991

greg0ire commented:
@aLkRicha what do you mean by reused? This migrations needs to be done on all bundles, right?

aLkRicha commented:
@greg0ire if we are using templating service in more then one sonata bundle, will be correct to create own templating service in base bundle (sonata core) and reuse it in block bundle, admin bundle, etc..

jordisala1991 commented:
One question , it is BC break to directly go to the step where we use twig directly? Without removing any code just deprecating what it is needed

covex-nn commented:
@jordisala1991 the second argument of [`Sonata\AdminBundle\Block\AdminListBlockService`](https://github.com/sonata-project/SonataBlockBundle/blob/3.x/src/Block/Service/AbstractBlockService.php) (abstract parent class for all Blocks) is `Symfony\Bundle\FrameworkBundle\Templating\EngineInterface $templating`.

So, we cannot just change `EngineInterface` to `Twig\Environment` in 3.x branch - this is BC break. And there is nothing to deprecate here, except may be `AbstractBlockService` itself.

aLkRicha commented:
@covex-nn please, answer my question ☺️  https://github.com/sonata-project/SonataBlockBundle/pull/457#issuecomment-356382537

covex-nn commented:
@aLkRicha CoreBundle does not use `templating` service, BlockBundle does. The main purpose
of `sonata.templating` is to replace `templating` as a second argument of `AdminListBlockService` class in 3.x branch without BC breaks. If some other bundle uses `templating`, then there could be another way how to decouple from `symfony/templating` and how to use `twig` service instead of `templating`.

So, BlockBundle is a right place for a new `sonata.templating` service. May be `sonata.templating` should be in separate repository (not in SonataCoreBundle and not in SonataBlockBundle), but i am not sure about it (it is too complicated)

jordisala1991 commented:
@covex-nn but you can remove the type hint, and check if the instance passed if from EngineInterface or TwigEnv.. and if not throw an exception.. and that is not BC break

covex-nn commented:
@jordisala1991 i used http://symfony.com/doc/current/contributing/code/bc.html to determine what change is allowed and what is not. Removing type hint of an argument of a public method is not allowed there. I thought it is a BC break.

Anyway, removing type hint could be done only after all template references in all bundles will be guaranteed switched from "_templating syntax_" to "_twig namespaced syntax_", i.e. only in master branch, in next major release. But i hoped to run SonataAdminBundle with Flex with in 3.x branch.

If "_removing type hint of an argument_" is not BC break, than my plan is not good enough, and there will be 2 stages, not 3.

jordisala1991 commented:
Wow, they really consider it bc break, althought we have done it several times, and I am not sure why they consider a bc break, because it does not break anything (if done correctly) . Wdyt @greg0ire?

greg0ire commented:
Removing the type hint of a method is a BC break for extending classes (they may only widen it).

jordisala1991 commented:
https://3v4l.org/s6f92

It throws a warning but it works, anyway if we consider it BC break we cannot do what I said.

greg0ire commented:
Warning are BC breaks... Even strict standards are. That's why classes / methods should be final

covex-nn commented:
I've updated issue, i added removing type hint for second argument of `Sonata\AdminBundle\Block\AdminListBlockService` in current master. Both `sonata.templating` and `twig` services will be used in blocks. Future master branch will not contain `sonata.templating` service.

greg0ire commented:
The build fails on php 7.0

OskarStark commented:
> The build fails on php 7.0

I restarted the build, as it was a race condition 👍

aLkRicha commented:
do we really need there `symfony/templating` dep?

covex-nn commented:
@aLkRicha we need it in 3.x branch, because Templating Component is still used. It will be removed in master branch, if this PR will be accepted

jordisala1991 commented on src/Test/FakeTemplating.php:
```diff
@@ -11,7 +11,7 @@
 namespace Sonata\BlockBundle\Test;
 
-use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+use Sonata\BlockBundle\Templating\EngineInterface;
```
This is the only “use” that you are replackng even though there are more referencing to EngineInterface on the project. Dont we need to replace all of them?

covex-nn replied:
We will replace all "use"s in master branch, if this PR will be accepted. Tests, that use `Sonata\BlockBundle\Test\FakeTemplating` show us, that when we use templating service, implementing `Sonata\BlockBundle\Templating\EngineInterface`, everything is fine

covex-nn replied:
And we use `sonata.templating` instead of `templating` !

`sonata.templating` is `Sonata\BlockBundle\Templating\TwigEngine`, it implements `Sonata\BlockBundle\Templating\EngineInterface`.

jordisala1991 commented:
Lets pull the trigger, great work @covex-nn

covex-nn commented:
@jordisala1991 please do not release it right now. I realized, that i should not remove tag from templating helper. I will create PR to fix that
